File: Maquina-Turing/Machine.cs
using System.Text.RegularExpressions;

class Machine
{
    public string input {get; set;}
    public string[] states {get; set;}
    public string[] alphabet {get; set;}
    public string[] alphabet_pile {get; set;}
    public string blanc_symbol {get; set;}
    public string acceptances_state {get; set;}
    public string reject_state {get; set;}
	public string initial_state {get; set;}
	public Dictionary<string, string> transitions {get; set;}
	public List<string> configurations {get; set;}

    public Machine(string Input, string[] States, string[] Alphabet, 
		    string[] Alphabet_pile, string Blanc_symbol, 
		    string Acceptances_state, string Reject_state,
			string Initial_state, Dictionary<string, string> Transitions) {
	input = Input;
	states = States;
	alphabet = Alphabet;
	alphabet_pile = Alphabet_pile;
	blanc_symbol = Blanc_symbol;
	acceptances_state = Acceptances_state;
	reject_state = Reject_state;
	initial_state = Initial_state;
	transitions = Transitions;
	configurations = [];
    }

    private List<string> tape_initialization(string input, string blanc) 
    {
    	List<string> input_tape = new List<string> { blanc, blanc };
	for (int i = 0; i < input.Length; i++)
	{
		input_tape.Add(Convert.ToString(input[i]));
	}

	input_tape.Add(blanc);
	input_tape.Add(blanc);

	return input_tape;
    }

	public bool derivation() 
	{
		List<string> tape = tape_initialization(input, blanc_symbol);
		string state = initial_state;
		int index = 2;
		string char_tape = tape[index];

		while (!state.Equals(acceptances_state) && !state.Equals(reject_state))
		{
			string value = transition(state, char_tape);
			string[] value_list = value.Split(",");
			
			state = value_list[0]; //Change old state for new one
			tape[index] = value_list[1]; //Change the char in the tape
			if (value_list[2].Equals("R"))
			{
				index += 1;
			}
			else
			{
				index -= 1;
			}

			save_config(string.Join("", tape), index, state);

			if (configurations.Count > 50) {
				configurations.Add("\n*****Loop*****\n");
				return false;
			}

			char_tape = tape[index]; //New char to read in tape
		}

		return state.Equals(acceptances_state);
	}

	private string transition(string state, string char_tape) {
		string key = state + ',' + char_tape;
		string value = transitions[key];

		return value;
	}

	public void generate_data()
	{
		var formattedTransitions = transitions.Select(kv => $"({kv.Key}) : ({kv.Value})");

		string output_path = "output.txt";
		string content = 
		"Q: "+ string.Join(",", states) + "\n" +
		"Σ: " + string.Join(",",alphabet) + "\n" +
		"Γ: " + string.Join(",",alphabet_pile) + "\n" +
		"q0: " + initial_state + "\n" +
		"Qaccept: " + acceptances_state + "\n" +
		"Qreject: " + reject_state + "\n" +
		"δ: \n" + "{\n" +
		string.Join("\n", formattedTransitions) + 
		"\n}\n\n" +
		"Configuraciones:\n\n" + string.Join("\n", configurations);

		File.WriteAllText(output_path, content);
	}

	private void save_config(string actualTape, int index, string state)
	{
		string str = Regex.Replace(actualTape, @"[\[\],\s]", "");
		string firstHalf = str.Substring(0, index - 1);
		string lastHalf = str.Substring(index, str.Length - index);

		configurations.Add(firstHalf + state + lastHalf);
	}

	
}

File: Machine.java
import java.util.*;

public class Machine {
    private String input;
    private String[] states;
    private String[] alphabet;
    private String[] alphabetPile;
    private String blancSymbol;
    private String acceptanceState;
    private String rejectState;
    private String initialState;
    private Map<String, String> transitions;
    private List<String> configurations;

    public Machine(String input, String[] states, String[] alphabet, String[] alphabetPile, String blancSymbol,
            String acceptanceState, String rejectState, String initialState, Map<String, String> transitions) {
        this.input = input;
        this.states = states;
        this.alphabet = alphabet;
        this.alphabetPile = alphabetPile;
        this.blancSymbol = blancSymbol;
        this.acceptanceState = acceptanceState;
        this.rejectState = rejectState;
        this.initialState = initialState;
        this.transitions = transitions;
        this.configurations = new ArrayList<String>();
    }

    private List<String> tapeInitialization(String input, String blanc) {
        List<String> inputTape = new ArrayList<>(Arrays.asList(blanc, blanc));
        for (int i = 0; i < input.length(); i++) {
            inputTape.add(String.valueOf(input.charAt(i)));
        }
        inputTape.add(blanc);
        inputTape.add(blanc);
        return inputTape;
    }

    public boolean derivation() {
        List<String> tape = tapeInitialization(input, blancSymbol);
        String state = initialState;
        int index = 2;
        String charTape = tape.get(index);

        while (!state.equals(acceptanceState) && !state.equals(rejectState)) {
            String value = transition(state, charTape);
            String[] valueList = value.split(",");

            state = valueList[0]; // Change old state for new one
            tape.set(index, valueList[1]); // Change the char in the tape
            if (valueList[2].equals("R")) {
                index += 1;
            } else {
                index -= 1;
            }

            saveConfig(tape.toString(), index, state);

            if (configurations.size() > 50) {
                configurations.add("\n*****Loop*****\n");
                return false;
            }

            charTape = tape.get(index); // New char to read in tape
        }

        return state.equals(acceptanceState);
    }

    private String transition(String state, String charTape) {
        String key = state + "," + charTape;
        return transitions.get(key);
    }

    public void generateData(boolean result) {
        String outputPath = "output.txt";
        String content = configurations.toString();
        content = content.replaceAll("[\\[\\] \\s]", "") // Elimina '[' y ']'
                .replace(",", "\n"); // Reemplaza las comas por saltos de línea

        if (result) {
            content += "\n\nCadena aceptada";
        } else {
            content += "\n\nCadena rechazada";
        }

        try {
            java.nio.file.Files.write(java.nio.file.Paths.get(outputPath), content.getBytes());
        } catch (java.io.IOException e) {
            e.printStackTrace();
        }
    }

    private void saveConfig(String actualTape, int index, String state) {
        String str = actualTape.replaceAll("[\\[\\],\\s]", "");
        String firstHalf = str.substring(0, index - 1);
        String lastHalf = str.substring(index, str.length() - 1);

        configurations.add(firstHalf + state + lastHalf);
    }
}
